src/main/java/edu/harvard/iq/dataverse/util/cache/RateLimitUtil.java

```diff
@@ -0,0 +1,140 @@
+        // Get the current number of tokens in the bucket
+        long tokens = longFromKey(rateLimitCache, key);
+        long tokensToAdd = (long) (deltaTime * tokensPerMinute);
+        if (tokensToAdd > 0) { // Don't update timestamp if we aren't adding any tokens to the bucket
+            tokens = min(capacityPerHour, tokens + tokensToAdd);
+            rateLimitCache.put(keyLastUpdate, String.valueOf(currentTime));
+        }
+        // Update with any added tokens and decrement 1 token for this call if not rate limited (0 tokens)
+        rateLimitCache.put(key, String.valueOf(max(0, tokens-1)));
+        return tokens < 1;
+    }
+
+    static int getCapacityByTierAndAction(SystemConfig systemConfig, Integer tier, String action) {
+        if (rateLimits.isEmpty()) {
+            init(systemConfig);
+        }
+
+        return rateLimitMap.containsKey(getMapKey(tier,action)) ? rateLimitMap.get(getMapKey(tier,action)) :
+                rateLimitMap.containsKey(getMapKey(tier)) ? rateLimitMap.get(getMapKey(tier)) :
+                        getCapacityByTier(systemConfig, tier);
+    }
+    static int getCapacityByTier(SystemConfig systemConfig, int tier) {
+        int value = NO_LIMIT;
+        String csvString = systemConfig.getRateLimitingDefaultCapacityTiers();
+        try {
+            if (!csvString.isEmpty()) {
+                int[] values = Arrays.stream(csvString.split(",")).mapToInt(Integer::parseInt).toArray();
+                if (tier < values.length) {
+                    value = values[tier];
+                }
+            }
+        } catch (NumberFormatException nfe) {
+            logger.warning(nfe.getMessage());
+        }
+        return value;
+    }
+    static void init(SystemConfig systemConfig) {
+        getRateLimitsFromJson(systemConfig);
+        /* Convert the List of Rate Limit Settings containing a list of Actions to a fast lookup Map where the key is:
+             for default if no action defined: "{tier}:" and the value is the default limit for the tier
+             for each action: "{tier}:{action}" and the value is the limit defined in the setting
+        */
+        rateLimitMap.clear();
+        rateLimits.forEach(r -> {
+            r.setDefaultLimit(getCapacityByTier(systemConfig, r.getTier()));
+            rateLimitMap.put(getMapKey(r.getTier()), r.getDefaultLimitPerHour());
+            r.getActions().forEach(a -> rateLimitMap.put(getMapKey(r.getTier(), a), r.getLimitPerHour()));
+        });
+    }
+    static void getRateLimitsFromJson(SystemConfig systemConfig) {
+        String setting = systemConfig.getRateLimitsJson();
+        rateLimits.clear();
+        if (!setting.isEmpty()) {
+            try {
+                JsonReader jr = Json.createReader(new StringReader(setting));
+                JsonObject obj= jr.readObject();
+                JsonArray lst = obj.getJsonArray("rateLimits");
+                rateLimits.addAll(gson.fromJson(String.valueOf(lst),
```

Comment:
changed

Reply:
I agree that it would be best to not introduce another dependency on Gson, which we'd like to remove some day:

- #6810

Jakarta EE comes with jakarta.json.Json and it would be best to use it.